examples/megssm/models.py

```diff
@@ -0,0 +1,872 @@
+        self._penalty = penalty
+
+        # initialize lists of smoothed estimates
+        self._mus_smooth_lst = None
+        self._sigmas_smooth_lst = None
+        self._sigmas_tnt_smooth_lst = None
+        self._loglik = None
+        self._store_St = bool(store_St)
+
+        # initialize sufficient statistics
+        timepts, num_roi, _ = self.A.shape
+        self._B0 = np.zeros((num_roi, num_roi))
+        self._B1 = np.zeros((timepts-1, num_roi, num_roi))
+        self._B3 = np.zeros((timepts-1, num_roi, num_roi))
+        self._B2 = np.zeros((timepts-1, num_roi, num_roi))
+        self._B4 = list()
+        
+        self._subject_data = dict()
+        
+    def add_subject(self, subject, subject_dir, epochs,labels, fwd, cov):
+        roi_to_src = ROIToSourceMap(fwd, labels) # compute ROI-to-source map
+        scales = {'eeg_scale' : 1, 'mag_scale' : 1, 'grad_scale' : 1} #scale=1 has no effect
+        fwd_src_snsr, fwd_roi_snsr, snsr_cov, epochs = \
+                _scale_sensor_data(epochs, fwd, cov, roi_to_src, **scales)
+                
+        # which_roi = roi_to_src.which_roi # array of len(sources); val = ROI of source
+        # data = epochs._data
+        # data = np.swapaxes(data,-1,-2)
+        # subjectdata = [(data, fwd_roi_snsr, fwd_src_snsr, snsr_cov, which_roi)]
+        
+        sdata = run_pca_on_subject(subject,epochs, fwd, cov, labels)
+        data, fwd_roi_snsr, fwd_src_snsr, snsr_cov, which_roi = sdata
+        subjectdata = [(data, fwd_roi_snsr, fwd_src_snsr, snsr_cov, which_roi)]
+        
+        self.set_data(subjectdata)
+        
+        # epochs, fwd_roi_snsr, fwd_src_snsr, snsr_cov, which_roi = subjectdata
+        self._subject_data[subject] = dict()
+        self._subject_data[subject]['epochs'] = epochs
+        self._subject_data[subject]['fwd_src_snsr'] = fwd_src_snsr
+        self._subject_data[subject]['fwd_roi_snsr'] = fwd_roi_snsr
+        self._subject_data[subject]['snsr_cov'] = snsr_cov
+        self._subject_data[subject]['labels'] = labels
+        self._subject_data[subject]['which_roi'] = which_roi
+
+    def set_data(self, subjectdata):
+        # add subject data, re-generate log_sigsq_lst if necessary
+        super().set_data(subjectdata)
+        if len(self.log_sigsq_lst) != self._nsubjects:
+            num_roi = self.log_sigsq_lst[0].shape[0]
+            self.log_sigsq_lst = [np.log(np.random.gamma(2, 1, size=num_roi))
+                                  for _ in range(self._nsubjects)]
+
+        # reset smoothed estimates and log-likelihood (no longer valid if
+        # new data was added)
+        self._mus_smooth_lst = None
+        self._sigmas_smooth_lst = None
+        self._sigmas_tnt_smooth_lst = None
+        self._loglik = None
+        self._B4 = [None] * self._nsubjects
+
+    # TODO: figure out how to initialize smoothed parameters so this doesn't
+    #       break, e.g. if em_objective is called before em for some reason
+    def em_objective(self):
```

Comment:
This should be private (prefixed by `_`) to make it clear users should never touch it